### DISCUSSION TITLE
How to use these input files: modules/external_petsc_solver/test/tests/...?

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#29164] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

How to use these input files: `modules/external_petsc_solver/test/tests/moose_as_parent.i; petsc_transient_as_sub.i; petsc_transient_as_parent.i; moose_as_sub.i` ? 
Recently, i'm trying to coupling external codes in MOOSE. Although https://github.com/idaholab/moose/discussions/28287 suggests that maybe choosing `external_petsc_solver` to study is a good choice because of its hard-coded, I'm still interested in it. At least I want them to run in my PC.

My steps are as follows and I encounter errors:
1. Create an Application like https://mooseframework.inl.gov/getting_started/new_users.html
```
cd ~/projects
./moose/scripts/stork.sh thmbryton
```
2. Edit `Makefile` with yes for `ALL_MODULES`
```
ALL_MODULES               := yes
```
3. Compile it in application directory
```
conda activate moose
make
```
4. Copy and paste `modules/external_petsc_solver/test/tests/moose_as_parent.i; petsc_transient_as_sub.i; petsc_transient_as_parent.i; moose_as_sub.i` into the application directory
5. Run MOOSE opt with *.i
```
./thmbryton-opt -i moose_as_parent.i
```
And it returns errors:
```
"ExternalPetscSolverApp" is not a registered application name.
Unable to locate library archive for "ExternalPetscSolverApp".
We attempted to locate the library archive "libexternal_petsc_solver-opt.la" in the following paths:
      ../lib
Make sure you have compiled the library and either set the "library_path" variable in your input file or exported "MOOSE_LIBRARY_PATH".
```
First, it says: `ExternalPetscSolverApp" is not a registered application name.`, I don't know what to do with it... Even I don't know why there is a `app_type`  parameter in  `TransientMultiApp` class https://mooseframework.inl.gov/source/multiapps/TransientMultiApp.html  . I think `TransientMultiApp` class is from parent file 
to drive sub file. Why does it need `app_type`?

**Comment by GiudGiud** [comment]
Hello

Can you you the executable in moose/modules/external_petsc_solver ? If you can build one

> why there is a app_type parameter in TransientMultiApp class

The app_type lets us build smaller apps, or different apps from dynamically loaded libraries, than the current app you are using for the main solve
It is generally not required. You can use the same app for the parent and child solves most often. The only times it is required is if the child input can only be understood/ran properly by a different app

Guillaume

**Reply by Wang-Yihu** [reply, marked as answer]
You're right sir. 
I input in the terminal in `modules/external_petsc_solver/`
```
conda activate moose
make
```
And then use this -opt in `modules/external_petsc_solver/test/tests/`
```
../../../external_petsc_solver-opt -i moose_as_parent.i
```
It really works!
You mean that `app_type` can make `*.i` files from different moose applications work together?

**Reply by GiudGiud** [reply]
Ok so by default we do not include the external_petsc_solver module in anyone s app build. I don’t think it s ever needed since it’s just an example. 

Yes using app_type you can run inputs of a different app in a multiapp

**Reply by Wang-Yihu** [reply]
OK, thanks. Since I do not want to run a `*.i` file in another app at least for now, I do not want to implement this function. If I want to couple external codes in MOOSE, is making out codes in external_petsc_solver a good idea? I've mentioned that there're only five parts: base, mesh, petscsolver, problems, and timesteppers in this module, though there are some code from petsc I cannot understand now... 
I'm trying to understand cardinal, https://github.com/neams-th-coe/cardinal , but there are too many codes...

**Reply by GiudGiud** [reply]
Cardinal is a better example than the external_petsc_solver of coupling external codes into MOOSE
